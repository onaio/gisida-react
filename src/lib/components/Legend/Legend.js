/* eslint-disable no-loop-func */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Actions, formatNum, hexToRgbA } from 'gisida';
import { buildLayersObj } from '../../utils';
import Parser from 'html-react-parser';
import './Legend.scss';

const mapStateToProps = (state, ownProps) => {
  const MAP = state[ownProps.mapId] || { layers: {}}
  return {
    layerObj: MAP.layers[MAP.activeLayerId],
    lastLayerSelected: MAP.layers[MAP.lastLayerSelected],
    layersData: buildLayersObj(MAP.layers),
    MAP,
    mapId: ownProps.mapId,
    primaryLayer: MAP.primaryLayer,
    showFilterPanel: MAP.showFilterPanel,
  }
}

export class Legend extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      setPrimary: false,
    };
  }

  onUpdatePrimaryLayer(e) {
    e.preventDefault();
    const { dispatch, mapId } = this.props;
    const targetLayer = e.currentTarget.getAttribute('data-layer');
    // dispatch primary layer id
    dispatch(Actions.updatePrimaryLayer(mapId, targetLayer));
  }

  render() {
    const { layerObj, mapId, lastLayerSelected } = this.props;

    if (!layerObj) {
      return false;
    }

    const legendItems = [];

    let primaryLegend;
    let layer;
    for (let l = 0; l < this.props.layersData.length; l += 1) {
      layer = this.props.layersData[l];
      const circleLayerType = (layer && layer.credit && layer.type === 'circle' && !layer.categories.shape && layer.visible);
      const symbolLayer = (layer && layer.credit && layer.categories &&  layer.categories.shape && layer.type !== 'circle'); 
      const fillLayerNoBreaks = (layer && layer.credit && layer.categories && layer.categories.breaks === 'no');
      const fillLayerWithBreaks = (layer && layer.credit && layer.type !== 'chart' && layer.type !== 'circle' && layer.categories && layer.categories.breaks === 'yes');
      const activeLayerSelected =  this.props.primaryLayer === layer.id ? 'primary' : '';

      let background = [];

      let uniqueStops;

      const quantiles = [];

      if (circleLayerType && layer.breaks && layer.stopsData && layer.styleSpec && layer.styleSpec.paint) {
        const stopVals = [];
        layer.stopsData.forEach((s) => {
           stopVals.push(s[1]);
        });

        layer.styleSpec.paint['circle-radius'].stops.forEach((s) => {
          stopVals.push(s[1]);
        });

        uniqueStops = [...new Set(stopVals)].sort((a, b) => a - b);

        uniqueStops.forEach((s) => {
          quantiles.push((
            <span
              className="circle-container"
              key={s}>
              <span
                style={
                  {
                    background: Array.isArray(layer.categories.color) ? layer.categories.color[uniqueStops.indexOf(s)]
                      : layer.categories.color,
                    width: `${s * 2}px`,
                    height: `${s * 2}px`,
                    margin: `0px ${uniqueStops.indexOf(s) + 2}px`
                  }
                }
              ></span>
              <p>{layer.breaks[uniqueStops.indexOf(s)]}</p>
              </span>
          ));
        });
      }

      if (lastLayerSelected && lastLayerSelected.id === layer.id) {
        if (circleLayerType) {
          primaryLegend = (
            <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-shapes legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <div className="legend-symbols">
              {quantiles}
            </div>
            <span>{layer.credit}</span>
          </div>);
        }
        if (fillLayerNoBreaks && !layer.parent) {
          const fillWidth = (100 / layer.categories.color.filter(c =>
            c !== "transparent").length).toString();

          layer.categories.color.forEach((color, index) => {
            if (color !== "transparent") {
              background.push((
                <li
                  key={index}
                  style={{ background: color, width: `${fillWidth}%` }}>
                  {layer.categories.label[index]}
                </li>
              ));
            }
          });

          const legendClass = layer.categories ? 'legend-label' : '';

          primaryLegend = (
            <div
              id={`legend-${layer.id}-${mapId}`}
              className={`legend-row ${activeLayerSelected}`}
              data-layer={`${layer.id}`}
              onClick={(e) => this.onUpdatePrimaryLayer(e)}
              key={l}
            >
              <b>
                {layer.label}
              </b>
              <div className={`legend-fill ${legendClass}`}>
                <ul>
                  {background}
                </ul>
              </div>
              <span>
                {Parser(layer.credit)}
              </span>
            </div>
          );
        } if (fillLayerWithBreaks && layer.stops && !layer.parent) {
          const { stopsData, breaks, colors, Data } = layer;

          const dataValues = Data.map(values => parseInt(values[layer.property], 10));
          const colorLegend = [...new Set(stopsData.map(stop => stop[1]))];
          const legendSuffix = layer.categories.suffix ? layer.categories.suffix : '';

          if (colorLegend.includes('transparent') && !(colors).includes('transparent')) {
            colors.splice(0, 0, 'transparent');
            breaks.splice(1, 0, breaks[0]);
          }

          colorLegend.forEach((color, index) => {
            const stopsIndex = layerObj.stops[4].indexOf(color);

            if (stopsIndex !== -1) {
              const firstVal = stopsIndex ? layerObj.stops[3][stopsIndex - 1] : 0;
              const lastVal = layerObj.stops[3][stopsIndex];
              background.push((
                <li
                  key={index}
                  className={`background-block-${layer.id}-${mapId}`}
                  data-tooltip={`${formatNum(firstVal, 1)}-${formatNum(lastVal, 1)}${legendSuffix}`}
                  style={{ background: hexToRgbA(color, 0.9).toString(), width: (100 / colorLegend.length) + '%' }}
                >
                </li>
              ));
            }
          });

          primaryLegend = (
            <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <ul
              className="legend-limit"
              style={{ padding: '0% 0% 3% 0%' }}
            >
              <li
                id={`first-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', left: '3%' }}
              >
                {0}
                {legendSuffix}
              </li>
              <li
                id={`last-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', right: '3%' }}
              >
                {formatNum(Math.max(...dataValues), 1)}
                {legendSuffix}
              </li>
            </ul>
            <div
              className="legend-fill"
            >
              <ul
                id="legend-background"
              >
                {background}
              </ul>
            </div>
            <span>{Parser(layer.credit)}</span>
          </div>
          );

        }
        continue;
      }
      if (circleLayerType) {
        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-shapes legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <div className="legend-symbols">
              {quantiles}
            </div>
            <span>{layer.credit}</span>
          </div>  
        ));
      } else if (symbolLayer) {
        layer.categories.color.forEach((color, index) => {
          const style = layer.categories.shape[index] === 'triangle-stroked-11' ||
          layer.categories.shape[index] === 'triangle-15' ?
          'border-bottom-color:' : 'background:';
          const styleString = `${style}: ${color}`;
          background += (
            <li
              className="layer-symbols"
              key={index}
            >
              <span
                className={`${layer.categories.shape[index]}`}
                style={{styleString}}
              />
              {layer.categories.label[index]}
            </li>
          );
        });

        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
            key={l}
          >
            <b>
              {layer.label}
            </b>
            <div className="legend-shapes">
              <ul style={{ left: '0' }}>
                {background}
              </ul>
            </div>
            <span>
              {Parser(layer.credit)}
            </span>
          </div>
        ));

      } else if (fillLayerNoBreaks && !layer.parent) {
        const fillWidth = (100 / layer.categories.color.filter(c =>
          c !== "transparent").length).toString();

        layer.categories.color.forEach((color, index) => {
          if (color !== "transparent") {
            background.push((
              <li
                key={index}
                style={{ background: color, width: `${fillWidth}%` }}>
                {layer.categories.label[index]}
              </li>
            ));
          }
        });

        const legendClass = layer.categories ? 'legend-label' : '';

        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
            key={l}
          >
            <b>
              {layer.label}
            </b>
            <div className={`legend-fill ${legendClass}`}>
              <ul>
                {background}
              </ul>
            </div>
            <span>
            {Parser(layer.credit)}
            </span>
          </div>
        ));
      } else if (fillLayerWithBreaks && layer.stops && !layer.parent) {
        const { stopsData, breaks, colors, Data } = layer;
        const dataValues = Data.map(values => parseInt(values[layer.property], 10));
        const colorLegend = [...new Set(stopsData.map(stop => stop[1]))];
        const legendSuffix = layer.categories.suffix ? layer.categories.suffix : '';

          if (colorLegend.includes('transparent') && !(colors).includes('transparent')) {
            colors.splice(0, 0, 'transparent');
            breaks.splice(1, 0, breaks[0]);
          }

          colorLegend.forEach((color, index) => {
            const stopsIndex = layerObj.stops[4].indexOf(color);

            if (stopsIndex !== -1) {
              const firstVal = stopsIndex ? layerObj.stops[3][stopsIndex - 1] : 0;
              const lastVal = layerObj.stops[3][stopsIndex];
              background.push((
                <li
                  key={index}
                  className={`background-block-${layer.id}-${mapId}`}
                  data-tooltip={`${formatNum(firstVal, 1)}-${formatNum(lastVal, 1)}${legendSuffix}`}
                  style={{ background: hexToRgbA(color, 0.9).toString(), width: (100 / colorLegend.length) + '%' }}
                >
                </li>
              ));
            }
          });

        legendItems.unshift((
          <div
            id={`legend-${layer.id}-${mapId}`}
            className={`legend-row ${activeLayerSelected}`}
            data-layer={`${layer.id}`}
            key={l}
            onClick={(e) => this.onUpdatePrimaryLayer(e)}
          >
            <b>
              {layer.label}
            </b>
            <ul
              className="legend-limit"
              style={{ padding: '0% 0% 3% 0%' }}
            >
              <li
                id={`first-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', left: '3%' }}
              >
                {0}
                {legendSuffix}
              </li>
              <li
                id={`last-limit-${layer.id}`}
                className={`${mapId}`}
                style={{ position: 'absolute', listStyle: 'none', display: 'inline', right: '3%' }}
              >
                {formatNum(Math.max(...dataValues), 1)}
                {legendSuffix}
              </li>
            </ul>
            <div
              className="legend-fill"
            >
              <ul
                id="legend-background"
              >
                {background}
              </ul>
            </div>
            <span>{Parser(layer.credit)}</span>
          </div>
        ));
      }
      else {

      }
    }

    legendItems.unshift(primaryLegend);

    return (
      <div>
        <div
          className={`legend ${mapId}`}
          style={{ right: this.props.showFilterPanel ? '30px' : '20px' }}>
          {legendItems}
        </div>
      </div>
    );
  }
}

Legend.propTypes = {
  layerObj: PropTypes.objectOf(PropTypes.any),
  layersData: PropTypes.arrayOf(PropTypes.any).isRequired,
  MAP: PropTypes.objectOf(PropTypes.any).isRequired,
  primaryLayer: PropTypes.string.isRequired,
};

export default connect(mapStateToProps)(Legend);
