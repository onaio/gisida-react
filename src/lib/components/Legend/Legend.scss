.legend {
  position: absolute;
  bottom: 40px;
  right: 20px;
  width: 320px;
  background-color: #f7f6ef;
  z-index: 1000;
  color: #555;
  font-size: 11px;
  font-family: 'Montserrat', sans-serif;
  line-height: 1.2;

  .legend-row {
    padding: 15px;
    position: relative;
    box-shadow: 0 1px 7px rgba(0, 0, 0, 0.2);

    b {
      display: inline-block;
      padding-right: 14px;
    }

    & > span {
      padding-right: 30px;
      display: inline-block;
    }
  }

  [data-visible="false"] {
    display: none;
  }
}

.circle-legend {
  background-color: #f7f6ef;
  border-radius: 3px;
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.1);
  font: 12px/20px 'Helvetica Neue', Arial, Helvetica, sans-serif;
  padding: 10px;
  position: absolute;
  min-width: 150px;
  z-index: 1;

  h4 {
    margin: 0 0 10px;
    display: block !important;
  }

  p {
    margin-left: 40px;
    position: absolute;
    display: block;
    top: 0;
  }

  div {
    position: relative;

    span {
      border-radius: 50%;
      display: inline-block;
      margin-right: 5px;
      opacity: 0.8;
    }
  }
}

.mapboxgl-ctrl-attrib {
  font-size: 11px;
  padding: 3px !important;
}

.legend-fill {
  height: 10px;
  background: #eee;
  margin: 3px 0;

  ul {
    padding: 0;
    list-style-type: none;
  }

  li {
    width: 30px;
    float: left;
    display: block;
    height: 10px;
    color: #fff;
    opacity: 0.8;
  }
}

.legend-fill.legend-label {
  height: 17px;

  li {
    text-align: center;
    height: 17px;
    padding-top: 2px;
  }
}

.legend-shapes {
  position: relative;

  .legend-symbols {
    min-width: 320px;

    .circle-container {
      display: inline-grid;
      min-width: 5px;
      min-height: 5px;
      position: relative;
      text-align: center;
      padding: 0 4px;

      span {
        display: inline-block;
        border-radius: 50%;
      }

      p {
        padding-top: 5px;
      }
    }
  }

  .circle-sm {
    height: 6px;
    width: 6px;
    margin: 0 10px 0 0;
    border-radius: 50%;
    background: #eee;
    display: inline-block;
  }

  .circle-md {
    height: 9px;
    width: 9px;
    margin: 0 10px 0 0;
    border-radius: 50%;
    background: #eee;
    display: inline-block;
  }

  .circle-lg {
    height: 12px;
    width: 12px;
    margin: 0 10px 0 0;
    border-radius: 50%;
    background: #eee;
    display: inline-block;
  }

  .triangle-stroked-11 {
    width: 0;
    height: 0;
    margin: 0 10px 0 0;
    border-left: 4px solid transparent;
    border-right: 4px solid transparent;
    border-bottom: 8px solid white;
    display: inline-block;
  }

  .triangle-15 {
    width: 0;
    height: 0;
    margin: 0 10px 0 0;
    border-left: 6px solid transparent;
    border-right: 6px solid transparent;
    border-bottom: 12px solid white;
    display: inline-block;
  }
}

.layer-symbols {
  list-style: none;
  margin-left: -3em;
  margin-top: 1em;
}

[data-tooltip] {
  position: relative;
  z-index: 2;
  cursor: pointer;
}

[data-tooltip]::before,
[data-tooltip]::after {
  visibility: hidden;
  -ms-filter: "progid:DXImageTransform.Microsoft.Alpha(Opacity=0)";
  filter: "progid: DXImageTransform.Microsoft.Alpha(Opacity=0)";
  opacity: 0;
  pointer-events: none;
}

[data-tooltip]::before {
  position: absolute;
  bottom: 5%;
  left: 50%;
  margin-bottom: 6px;
  margin-left: -82px;
  padding: 5px;
  width: 160px;
  -webkit-border-radius: 3px;
  -moz-border-radius: 3px;
  border-radius: 3px;
  background-color: transparent;
  color: #000;
  content: attr(data-tooltip);
  text-align: center;
  font-size: 12px;
  line-height: 1.2;
}

[data-tooltip]:hover::before,
[data-tooltip]:hover::after {
  visibility: visible;
  -ms-filter: "progid:DXImageTransform.Microsoft.Alpha(Opacity=100)";
  filter: "progid: DXImageTransform.Microsoft.Alpha(Opacity=100)";
  opacity: 1;
}

.swatch {
  display: inline-block;
  height: 10px;
  width: 5px;
}

@media screen and (max-width: 795px) {
  .ona-logo {
    display: none;
  }
}

.legend-row {
  &:not(.primary) {
    cursor: pointer;

    & > * {
      display: none;
    }

    b {
      display: inline-block;
      font-weight: normal;
      color: #999;
    }

    &:hover {
      box-shadow: 0 1px 7px rgba(0, 0, 0, 0.2), inset 0 0 0 1.5px rgba(0, 0, 0, 0.2);

      b {
        color: #555;
      }
    }
  }

  &.primary {
    box-shadow: 0 1px 7px rgba(0, 0, 0, 0.2), inset 0 0 0 1.5px rgba(0, 0, 0, 0.4);
    min-height: 100px;
  }
}