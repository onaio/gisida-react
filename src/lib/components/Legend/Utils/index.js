import React from 'react';
export const buildTimestamp = (layerObj) => {
    if (
        layerObj['timestamp'] &&
        Array.isArray(layerObj.source.data.features || layerObj.source.data) &&
        (layerObj.source.data.features || layerObj.source.data).length > 1
      ) {
        return (layerObj.source.data.features || layerObj.source.data)
          .map(d => (d.properties && d.properties[layerObj['timestamp']]) || d[layerObj['timestamp']])
          .sort()
          .reverse()[0];
      }
      return null;
}

export const buildDescription = (layerObj) => {
    return <div className='description'>
                <span className='glyphicon glyphicon-info-sign' />
                <p>{layerObj['legendDescription']}</p>
		   </div>;
}

export const combinedLinks = (legendDescription) => {
    if (legendDescription) {
        return (<div className='combinedLinks'>
                    {legendDescription}&nbsp;&nbsp;
                </div>);
            } else {
                return null
            }
}