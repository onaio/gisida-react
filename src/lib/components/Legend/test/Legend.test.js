import React from 'react';
import { Legend } from '../Legend';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Legend', () => {
  const layerObj = {"testlayer": "layer1"}
  const timeSeriesObj = {"testTimelayer": "timeLayer1"};
  const layers = {
    'test-layer-1': {
      'visible': true,
      'credit': "legend"
    }
  };
  const layersData = [layerObj];

  const componentWrapper = shallow(
    <Legend
      layerObj={layerObj}
      timeSeriesObj={timeSeriesObj}
      layersData={layersData}
      mapId='map-1'
      primaryLayer=''
      layers={layers}
      activeLayerIds={['test-layer-1']}
      MAP={{}}
    />
  );

  it('component renderes correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });
});
