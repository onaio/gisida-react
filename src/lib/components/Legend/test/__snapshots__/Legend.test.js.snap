// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend component renderes correctly 1`] = `
<div>
  <div
    className="legend map-1"
  />
</div>
`;
